### DISCUSSION TITLE
Adding a third option for controlling memory

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#220] [General]
A couple of weeks ago I read this article https://lwn.net/Articles/934679/ and reminded me it can be possible to emulate C++'s RAII behavior using GCC's / clang's extension via `cleanup` attribute.

@edubart what is your opinion on this kind of mechanism, assuming you are willing to give it a go and let users know this option is available only under GCC and clang?

It's just food for thought, nothing more.

**Comment by edubart** [comment]
I like this mechanism, and Nelua has official support for something like it. It's the `defer` statement and the `<close>` annotation of to-be-closed variables. I added a new section about to-be-closed variables in the overview, please check:

https://nelua.io/overview/#defer
https://nelua.io/overview/#to-be-closed-variables

**Reply by stefanos82** [reply]
Thank you for letting me know @edubart .

Yes, I know about `defer` and `<close>`; my question regarding these two, if I use any of those, does it mean a `## pragmas.nogc = true` is being used implicitly behind the scenes?

**Reply by stefanos82** [reply]
What I really meant was a mechanism to emit such code in C that helps with controlling memory scope.

**Reply by edubart** [reply]
> if I use any of those, does it mean a ## pragmas.nogc = true is being used implicitly behind the scenes?

Both defer statement and to-be-closed variables are orthogonal and unrelated to GC, that is, both work with or without GC, the GC has nothing to do with them.

> What I really meant was a mechanism to emit such code in C that helps with controlling memory scope.

The `cleanup` extension could be used to implement both defer and to-be-closed variables, however Nelua's C code generator is more explicit and explicitly emits the scope ending code. I think this way makes it more compatible with other C compilers.

**Comment by stefanos82** [comment]
> Both defer statement and to-be-closed variables are orthogonal and unrelated to GC, that is, both work with or without GC, the GC has nothing to do with them.
> 

Ah I see.

> The `cleanup` extension could be used to implement both defer and to-be-closed variables, however Nelua's C code generator is more explicit and explicitly emits the scope ending code. I think this way makes it more compatible with other C compilers.

Nice! That's good to know; I learned something new today!

Cheers @edubart :+1:
